shared/common-adapters/dropdown.native.js

```diff
@@ -0,0 +1,203 @@
+// @flow
+import React, {Component} from 'react'
+import {TouchableWithoutFeedback, Picker, Modal} from 'react-native'
+import Platform, {OS} from '../constants/platform'
+import {globalStyles, globalColors} from '../styles/style-guide'
+import Text from './text'
+import Box from './box'
+import Icon from './icon'
+import type {Props} from './dropdown'
+
+/*
+ * A dropdown on ios and android.
+ * The ios version will show a modal with an inline picker, clicking in the modal selects the option.
+ * The android version uses the built in picker but invisibly so it can show its modal correctly.
+ */
+
+// sentry for the 'other' value
+const otherItemValue = 'otherItemValue'
+// sentry for the 'pick' value
+const pickItemValue = 'pickItemValue'
+
+type State = {
+  modalVisible: boolean,
+  value: ?string
+}
+
+class Dropdown extends Component {
+  state: State;
+  showingPick: boolean;
+
+  constructor (props: Props) {
+    super(props)
+
+    this.showingPick = !this.props.value
+
+    this.state = {
+      modalVisible: false,
+      value: this._stateValue(this.props.value)
+    }
+  }
+
+  _stateValue (value: ?string): string {
+    return value || pickItemValue
+  }
+
+  componentWillReceiveProps (nextProps: Props) {
+    this.setState({value: this._stateValue(nextProps.value)})
+  }
+
+  shouldComponentUpdate (nextProps: Props, nextState: State): boolean {
+    if (
+      this.state.value !== nextState.value ||
+      this.state.modalVisible !== nextState.modalVisible ||
+      this.props.value !== nextProps.value) {
+      return true
+    }
+
+    return false
+  }
+
+  _selected () {
+    // Didn't actually select anything
+    if (this.state.value === pickItemValue) {
+      this.props.onClick(null, -1)
+    } else if (!this.state.value && this.props.onOther) {
+      this.props.onOther()
+    } else if (this.props.onClick) {
+      this.props.onClick(this.state.value, (this.props.options || []).indexOf(this.state.value))
+    }
+  }
+
+  _showModal (show: boolean) {
+    this.setState({modalVisible: show})
+
+    if (show) {
```

Comment:
I feel like this should be in different functions.

One for `_showModal` that just updates the modal state.

Another for setting the default value on open (e.g. `_maybeSetValueToFirstOption`)

and remove the call to `this._selected()` and doing that in the `onPress`, e.g. `<TouchableWithoutFeedback onPress={() => { this._showModal(false); this._selected()}}>`


Otherwise it is a bit confusing that this function does so much